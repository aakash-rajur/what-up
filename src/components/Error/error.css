.error-container {
    width: fit-content;
    max-width: 100%;
    display: inline-flex;
    flex-direction: column;
    align-items: center;
}

.error-container img {
    width: 52px;
    height: 52px;
}

.error-container span {
    text-align: center;
    font-size: 0.7em;
    font-weight: 100;
    letter-spacing: 0.2em;
}

.error-container pre {
    white-space: pre-wrap;
    text-align: center;
    font-weight: 100;
    line-height: 1.5em;
    font-size: 0.9em;
    padding: 0 1em;
}
