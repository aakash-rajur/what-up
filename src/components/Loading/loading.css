@import "../../utils/variables.css";

@keyframes spin {
    from {
        transform: rotate(0deg);
    }
    to {
        transform: rotate(360deg);
    }
}

.loading {
    display: inline-block;
    width: 52px;
    height: 52px;
    border-radius: 50%;
    border: 4px solid transparent;
    border-top-color: var(--accent-color);
    border-right-color: var(--accent-color);
    animation: spin 1s linear infinite;
}
