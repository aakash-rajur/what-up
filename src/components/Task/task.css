@import "../../utils/variables.css";

.task {
    min-height: 70px;
    border: none;
    padding: 0.9em 1em 0.4em 1em;
    font-size: 1.5em;
    font-weight: 100;
    border-radius: 4px;
    color: gray;
    box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
    transition: all 0.3s cubic-bezier(.25, .8, .25, 1);
    background: white;
    z-index: 0;

    display: grid;
    grid-template: 1fr/2em 1fr auto;
    align-items: center;
}

.status {
    border: 1px solid lightgray;
    border-radius: 50%;
    position: relative;
}

.status:after {
    content: "✔";
    position: absolute;
    left: 0;
    top: 0;
    font-size: 0.6em;
    width: var(--icon-size);
    height: var(--icon-size);
    display: flex;
    align-items: center;
    justify-content: center;
    color: limegreen;
    opacity: 0;
    transition: all 0.3s;
}

.cancel {
    visibility: hidden;
}

.task:hover:not([data-status="CANCELLED"]) > .cancel {
    visibility: initial;
}

.task[data-status="COMPLETED"] > .status:after {
    opacity: 1;
}

.task[data-status="COMPLETED"] > .description {
    text-decoration: line-through;
    text-decoration-color: var(--accent-opaque);
    opacity: 0.5;
}

.task[data-status="CANCELLED"] > .status:after {
    opacity: 1;
    content: "✗";
    color: gray;
}

.task[data-status="CANCELLED"] > .description {
    text-decoration: line-through;
    opacity: 0.5;
}

.editable, .description {
    font-weight: 100;
    font-size: 1em;
}

.editable {
    border: none;
    outline: none;
    color: gray;
}

.description {
    transition: all 0.3s;
    cursor: pointer;
}

.updated {
    font-size: 0.4em;
    grid-column: span 3;
    font-weight: 100;
    justify-self: end;
    color: lightgray;
}

@media (max-width: 600px) {
    .task {
        font-size: 1.2em;
    }
}
