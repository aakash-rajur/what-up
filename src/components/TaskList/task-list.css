.loading-container, .error-container, .no-tasks, .task-list {
    flex-grow: 1;
    flex-shrink: 1;
}

.loading-container {
    display: flex;
    height: 200px;
    justify-content: center;
    align-items: center;
}

.error-container {
    height: 200px;
    justify-content: center;
    align-items: center;
}

.no-tasks {
    height: 200px;
    font-weight: 100;
    display: flex;
    justify-content: center;
    align-items: center;
}

.task-list {
    display: grid;
    grid-template: auto/auto;
    grid-gap: 16px;
    justify-content: center;
    margin: 30px 0;
}
