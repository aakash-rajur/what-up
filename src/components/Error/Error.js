import PropTypes from "prop-types";
import React from "react";
import ERROR_ICON from "../../assets/alert.svg";
import "./error.css";

function Error({text, className}) {
  return (
    <div className={`error-container ${className || ""}`}>
      <img src={ERROR_ICON} alt={text} />
      <pre>{text}</pre>
    </div>
  );
}

Error.propTypes = {
  text: PropTypes.string,
  className: PropTypes.string
};

Error.defaultProps = {
  text: "Something Went Wrong"
};

export default Error;
