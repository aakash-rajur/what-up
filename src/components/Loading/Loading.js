import React from "react";
import "./loading.css";

function Loading({className}) {
  return <div className={`loading ${className || ""}`}>&nbsp;</div>;
}

export default Loading;
