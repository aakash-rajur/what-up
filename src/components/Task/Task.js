import PropTypes from 'prop-types';
import React, {Component} from 'react';
import CANCEL_ICON from '../../assets/close.svg';
import {MutableTask} from "../../utils/apollo";
import {TASK_CREATED, TASK_TITLE} from "../../utils/constants";
import {getFormattedTimestamp} from "../../utils/library";
import Input from "../Input/Input";
import './task.css';

export class Task extends Component {
	static propTypes = {
		description: PropTypes.string,
		status: PropTypes.string,
		id: PropTypes.string,
		updated: PropTypes.string
	};
	
	constructor(props) {
		super(props);
		this.onDescriptionChange = this.onDescriptionChange.bind(this);
		this.onSaveDescription = this.onSaveDescription.bind(this);
		this.state = {
			description: props.description,
			editable: false
		};
	}
	
	render() {
		let {
			status,
			updated,
			updateTask,
			cancelTask
		} = this.props, {
			description,
			editable
		} = this.state;
		return (
			<div className="app-width task" data-status={status}>
				<div key='status' className="icon status"
				     title={status === TASK_CREATED ?
					     'Set Complete' : 'Set Active'}
				     onClick={updateTask}>
					&nbsp;
				</div>
				{editable ? (
					<Input key='description' value={description}
					       className='editable' autoFocus={true}
					       onChange={this.onDescriptionChange}
					       onBlur={this.onEditChange(false)}
					       onValueReturn={this.onSaveDescription}/>
				) : (
					<div className="description" key='description'
					     title={TASK_TITLE} onDoubleClick={this.onEditChange(true)}>
						{description}
					</div>
				)}
				<img key='remove' src={CANCEL_ICON} alt="cancel-task"
				     className="icon cancel" title='Set Cancel'
				     onClick={cancelTask}/>
				<i className="updated">{getFormattedTimestamp(updated)}</i>
			</div>
		)
	}
	
	onEditChange(editable) {
		return () => {
			let {description} = this.props;
			this.setState({editable, description: description.trim()});
		};
	}
	
	onDescriptionChange({target: {value}}) {
		this.setState({description: value});
	}
	
	async onSaveDescription() {
		let {id, editTask} = this.props,
			{description} = this.state;
		await editTask({id, description});
		this.setState({editable: false});
	}
}


export default MutableTask(Task);
