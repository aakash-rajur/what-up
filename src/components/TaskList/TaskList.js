import PropTypes from 'prop-types';
import React, {Component} from 'react';
import {withQueryTasks} from "../../utils/apollo";
import {TASK_ALL} from "../../utils/constants";
import Error from "../Error/Error";
import Loading from "../Loading/Loading";
import Task from "../Task/Task";
import './task-list.css';

export class TaskList extends Component {
	static propTypes = {
		tasks: PropTypes.array,
		loading: PropTypes.bool,
		error: PropTypes.any,
		filter: PropTypes.string
	};

	state = {
		tasks: [],
		loading: true
	};

	componentDidUpdate(prevProps) {
		if (this.props.loading !== prevProps.loading) {
			this.setState({loading: this.props.loading && !this.state.tasks.length});
		}
		if (this.props.tasks !== prevProps.tasks && this.props.tasks)
			this.setState({tasks: this.props.tasks, loading: false});
	}

	render() {
		let {error, filter} = this.props, {
			loading, tasks
		} = this.state;
		return (
			error ? <Error text={error && (typeof error === "string" ? error : JSON.stringify(error))}/> :
				loading ? (<div className="loading-container"><Loading/></div>) :
					!tasks.length ? (
						<div className="no-tasks">
							{filter === TASK_ALL ? 'all your tasks show up here!' :
								`your ${filter.toLowerCase()} tasks show up here!`}
						</div>
					) : (
						<div className="task-list">
							{tasks.map((task, index) => <Task {...task} index={index} key={index}/>)}
						</div>
					)
		)
	}
}

export default withQueryTasks(TaskList);
