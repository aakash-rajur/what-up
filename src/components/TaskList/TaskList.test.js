import React from 'react';
import tasks from '../../../mock/tasks';
import {TASK_ALL, TASK_COMPLETED} from "../../utils/constants";
import Error from '../Error/Error';
import Loading from "../Loading/Loading";
import {Task} from "../Task/Task";
import {TaskList} from './TaskList';

describe('TaskList component when loading', () => {
	let props = {
		tasks: [],
		loading: true,
		error: null,
		filter: TASK_ALL
	}, component = shallow(<TaskList {...props}/>);
	
	it('dom integrity', checkDOM(component, [{
		selector: 'div.loading-container',
		children: [{
			selector: Loading,
			length: 1
		}]
	}]));
	
	it('state integrity', () => {
		expect(component.state('tasks').length)
			.toBe(0);
		expect(component.state('loading'))
			.toBe(true)
	});
	
	matchSnapshot(component)
});

describe('TaskList with error', () => {
	let props = {
		tasks: [],
		loading: false,
		error: {hello: 'world'},
		filter: TASK_ALL
	}, component = shallow(<TaskList {...props}/>);
	
	it('dom integrity', checkDOM(component, [{
		selector: Error,
		props: {
			text: JSON.stringify(props.error)
		}
	}]));
	
	matchSnapshot(component)
});

describe('TaskList component with empty List', () => {
	let props = {
		tasks: [],
		loading: true,
		error: null,
		filter: TASK_ALL
	}, component = shallow(<TaskList {...props}/>);
	
	component.setProps({tasks: []});
	
	it('dom integrity for filter ALL', checkDOM(component, [{
		selector: 'div.no-tasks',
		text: 'all your tasks show up here!'
	}]));
	
	it('should render empty list msg for completed', () => {
		component.setProps({filter: TASK_COMPLETED});
		
		checkDOM(component, [{
			selector: 'div.no-tasks',
			text: `your ${TASK_COMPLETED.toLowerCase()} tasks show up here!`
		}])();
	});
	
	it('state integrity', () => {
		expect(component.state('tasks').length)
			.toBe(0);
		expect(component.state('loading'))
			.toBe(false)
	});
	
	setImmediate(matchSnapshot, component);
});

describe('TaskList component with List', () => {
	let props = {
		tasks: [],
		loading: false,
		error: null,
		filter: TASK_ALL
	}, component = shallow(<TaskList {...props}/>);
	
	component.setProps({loading: false, tasks: props.tasks = [...tasks]});
	
	it('should render task-list', () => {
		expect(component.find('div.task-list').exists())
			.toBe(true);
		expect(component.children().length)
			.toBe(props.tasks.length);
	});
	
	setImmediate(it, 'dom integrity', checkDOM(component, [{
		selector: 'div.task-list',
		children: [{
			selector: Task,
			length: tasks.length
		}]
	}]));
	
	it('state integrity', () => {
		expect(component.state('tasks').length)
			.toBe(tasks.length);
		expect(component.state('loading'))
			.toBe(false)
	});
	
	setImmediate(matchSnapshot, component);
});
