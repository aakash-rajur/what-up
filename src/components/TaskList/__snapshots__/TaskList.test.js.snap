// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList component when loading match snapshot 1`] = `
<div
  className="loading-container"
>
  <Loading />
</div>
`;

exports[`TaskList component with List match snapshot with non-empty list 1`] = `
<div
  className="task-list"
>
  <Apollo(Apollo(Apollo(Task)))
    created="1533233094860"
    description="task1"
    id="1533233094860"
    index={0}
    key="0"
    status="COMPLETED"
    updated="1533233094860"
  />
  <Apollo(Apollo(Apollo(Task)))
    created="1533233143045"
    description="task2"
    id="1533233143045"
    index={1}
    key="1"
    status="CREATED"
    updated="1533233143045"
  />
</div>
`;

exports[`TaskList with error match snapshot 1`] = `
<Error
  text="{\\"hello\\":\\"world\\"}"
/>
`;
